File: prj6/MinHeap.py
class MinHeap:

    def __init__(self):
        self.heap_array = list()
        self.size = 0

    def insert(self, data):
        self.heap_array.append(data)
        size = len(self.heap_array) - 1
        while size > 0 and self.heap_array[size] < self.heap_array[self.parent(size)]:
            self.swap()
            size = self.parent(size)

    def swap(self):
        self.heap_array[self.size], self.heap_array[self.parent(
            self.size)] = self.heap_array[self.parent(self.size)], self.heap_array[self.size]

    def parent(self, index):
        return (index - 1) // 2

    def left_child(self, index):
        return (2 * index) + 1

    def right_child(self, index):
        return (2 * index) + 2

    def delete(self):
        pass

    def min_heap_order(self, index):
        left = self.left_child(index)
        right = self.right_child(index)
        smallest = index
        if left < len(self.heap_array) and self.heap_array[left] < self.heap_array[index]:
            smallest = left
        if right < len(self.heap_array) and self.heap_array[right] < self.heap_array[smallest]:
            smallest = right
        if smallest != index:
            self.heap_array[index], self.heap_array[smallest] = self.heap_array[smallest], self.heap_array[index]
            self.min_heap_order(smallest)

    def min_data(self):
        if len(self.heap_array) == 0:
            print("Heap is Empty!!")
        return self.heap_array[0]

    def min(self):
        if len(self.heap_array) == 0:
            print("Heap is Empty!!")
        min_val = self.heap_array[0]
        last_val = self.heap_array.pop()
        if len(self.heap_array) > 0:
            self.heap_array[0] = last_val
            self.min_heap_order(0)
        return min_val

    def display(self):
        print(self.heap_array)

File: prj6/minheap.cpp
#include <iostream>

using namespace std;

class MinHeap {
private:
    int* heapArray;
    int capacity;
    int size;

public:
    MinHeap(int cap) : capacity(cap), size(0) {
        heapArray = new int[capacity];
    }

    ~MinHeap() {
        delete[] heapArray;
    }

    void insert(int value) {
        if (size == capacity) {
            cout << "Heap overflow\n";
            return;
        }

        size++;
        int i = size - 1;
        heapArray[i] = value;

        
        while (i > 0 && heapArray[parent(i)] > heapArray[i]) {
            swap(heapArray[i], heapArray[parent(i)]);
            i = parent(i);
        }
    }

    int extractMin() {
        if (size <= 0) {
            cout << "Heap underflow\n";
            return -1; 
        }

        if (size == 1) {
            size--;
            return heapArray[0];
        }

        int root = heapArray[0];
        heapArray[0] = heapArray[size - 1];
        size--;

        
        heapifyDown(0);

        return root;
    }

    void printHeap() {
        for (int i = 0; i < size; i++) {
            cout << heapArray[i] << " ";
        }
        cout << "\n";
    }

private:
    int parent(int i) {
        return (i - 1) / 2;
    }

    void swap(int &a, int &b) {
        int temp = a;
        a = b;
        b = temp;
    }

    void heapifyDown(int i) {
        int minIndex = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        if (left < size && heapArray[left] < heapArray[minIndex]) {
            minIndex = left;
        }

        if (right < size && heapArray[right] < heapArray[minIndex]) {
            minIndex = right;
        }

        if (i != minIndex) {
            swap(heapArray[i], heapArray[minIndex]);
            heapifyDown(minIndex);
        }
    }
};

